x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/adapitiveallocations/AdaptiveAllocationsScalerService.java

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.adapitiveallocations;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.ClientHelper;
+import org.elasticsearch.xpack.core.ml.action.GetDeploymentStatsAction;
+import org.elasticsearch.xpack.core.ml.action.UpdateTrainedModelDeploymentAction;
+import org.elasticsearch.xpack.core.ml.inference.assignment.AssignmentStats;
+import org.elasticsearch.xpack.core.ml.inference.assignment.TrainedModelAssignment;
+import org.elasticsearch.xpack.core.ml.inference.assignment.TrainedModelAssignmentMetadata;
+import org.elasticsearch.xpack.ml.MachineLearning;
+import org.elasticsearch.xpack.ml.notifications.SystemAuditor;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class AdaptiveAllocationsScalerService implements ClusterStateListener {
+
+    record Stats(long successCount, long pendingCount, long failedCount, double inferenceTime) {
+
+        long requestCount() {
+            return successCount + pendingCount + failedCount;
+        }
+
+        double totalInferenceTime() {
+            return successCount * inferenceTime;
+        }
+
+        Stats add(Stats value) {
+            long newSuccessCount = successCount + value.successCount;
+            long newPendingCount = pendingCount + value.pendingCount;
+            long newFailedCount = failedCount + value.failedCount;
+            double newInferenceTime = newSuccessCount > 0
+                ? (totalInferenceTime() + value.totalInferenceTime()) / newSuccessCount
+                : Double.NaN;
+            return new Stats(newSuccessCount, newPendingCount, newFailedCount, newInferenceTime);
+        }
+
+        Stats sub(Stats value) {
+            long newSuccessCount = successCount - value.successCount;
+            long newPendingCount = pendingCount - value.pendingCount;
+            long newFailedCount = failedCount - value.failedCount;
```

Comment:
Should we handle wraps better? E.g. these should all be non-negative right so we could `max(..., 0)`. Also I know we trap this on `requestCount() > 0` but NaN runs the risk of polluting the estimates for ever more. Perhaps we should treat NaN as skip the update in the KalmanFilter.

Reply:
Added `max(0, ...)`. Also added `isNan` check just in case (though it shouldn't ever trigger)